src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,111 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] dataArray;
+    private int size;
+
+    public ArrayList() {
+        dataArray = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        growUp();
+        dataArray[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        growUp();
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index < size && index >= 0) {
+            System.arraycopy(dataArray, index, dataArray, index + 1, size - index);
+            dataArray[index] = value;
+            size++;
+            return;
+        }
+        throw new ArrayListIndexOutOfBoundsException(
+                "Index " + index + " is wrong for size " + size);
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (list.size() > size) {
+            growUp();
+        }
+        for (int i = 0; i < list.size(); i++) {
+            dataArray[size] = list.get(i);
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validationForIndex(index);
+        return (T)dataArray[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        validationForIndex(index);
+        growUp();
+        if (index == size) {
+            add(value);
+        }
+        dataArray[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validationForIndex(index);
+        Object elemetRemove = dataArray[index];
```

Comment:
```suggestion
        Object removedElement = dataArray[index];
```
better change to this name